2.0/m_checkbans.cpp

```diff
@@ -0,0 +1,263 @@
+	{
+		Channel* c = ServerInstance->FindChan(parameters[0]);
+		if (!c)
+		{
+			user->WriteNumeric(ERR_NOSUCHCHANNEL, "%s %s :No such channel", user->nick.c_str(), parameters[0].c_str());
+			return CMD_FAILURE;
+		}
+
+		// Only allow checking for matching users if you have access to the ban list
+		if (!CanCheck(c, user))
+			return CMD_FAILURE;
+
+		// Check for ban exceptions as mode 'e'
+		ListModeBase* ex = CheckExceptionMode();
+
+		// Loop through all users of the channel, checking for matches to bans and exceptions (if available)
+		const UserMembList* users = c->GetUsers();
+		for (UserMembCIter u = users->begin(); u != users->end(); u++)
+			CheckLists(user, c, u->first, ex);
+
+		user->WriteNumeric(RPL_ENDLIST, "%s %s :End of check bans list", user->nick.c_str(), c->name.c_str());
+		return CMD_SUCCESS;
+	}
+};
+
+class CommandTestBan : public Command
+{
+ public:
+	CommandTestBan(Module* Creator) : Command(Creator, "TESTBAN", 2, 2)
+	{
+		this->syntax = "<channel> <mask>";
+		this->Penalty = 6;
+	}
+
+	CmdResult Handle(const std::vector<std::string>& parameters, User* user)
+	{
+		Channel* c = ServerInstance->FindChan(parameters[0]);
+		if (!c)
+		{
+			user->WriteNumeric(ERR_NOSUCHCHANNEL, "%s %s :No such channel", user->nick.c_str(), parameters[0].c_str());
+			return CMD_FAILURE;
+		}
+
+		// Only allow testing bans if the user has access to set a ban on the channel
+		if (!CanCheck(c, user))
+			return CMD_FAILURE;
+
+		unsigned int matched = 0;
+		const UserMembList* users = c->GetUsers();
+		for (UserMembCIter u = users->begin(); u != users->end(); u++)
+		{
+			if (c->CheckBan(u->first, parameters[1]))
+			{
+				user->WriteNumeric(RPL_BANMATCH, "%s %s :Mask %s matches %s",
+					user->nick.c_str(), c->name.c_str(),
+					parameters[1].c_str(), u->first->nick.c_str());
+				matched++;
+			}
+		}
+
+		if (matched > 0)
+		{
+			float percent = ((float)matched / (float)users->size()) * 100;
+			user->WriteNumeric(RPL_BANMATCH, "%s %s :Mask %s matched %d of %lu users (%.2f%%).",
+				user->nick.c_str(), c->name.c_str(), parameters[1].c_str(),
+				matched, users->size(), percent);
+		}
+
+		user->WriteNumeric(RPL_ENDLIST, "%s %s :End of test ban list", user->nick.c_str(), c->name.c_str());
+		return CMD_SUCCESS;
+	}
+};
+
+class CommandWhyBan : public Command
+{
+ public:
+	CommandWhyBan(Module* Creator) : Command(Creator, "WHYBAN", 1, 2)
+	{
+		this->syntax = "<channel> [user]";
+	}
+
+	CmdResult Handle(const std::vector<std::string>& parameters, User* user)
+	{
+		Channel* c = ServerInstance->FindChan(parameters[0]);
+		if (!c)
+		{
+			user->WriteNumeric(ERR_NOSUCHCHANNEL, "%s %s :No such channel", user->nick.c_str(), parameters[0].c_str());
+			return CMD_FAILURE;
+		}
+
+		/* Allow checking yourself against channel bans with no access, but only
+		 * allow checking others if you have access to the channel ban list.
+		 */
+		User* u = parameters.size() == 1 ? user : NULL;
+		if (!u)
+		{
+			this->Penalty = 10;
```

Comment:
I believe what you meant here is `user->CommandFloodPenalty += 10000;` to set a user's penalty. This will currently modify the penalty for the command if anyone attempts to check another user.

Reply:
If I recall correctly, that's exactly my intention. The command handler will then assign the penalty.

Reply:
Yes but this currently sets the penalty for the command no matter what, after that line, any time the command is ran a penalty of 10 will be applied.

Reply:
Ah yeah I see what you're saying now. Thanks.